core/src/main/java/org/apache/calcite/plan/RelOptUtil.java

```diff
@@ -2199,13 +2199,49 @@ public static boolean equal(
       RelDataType type2,
       Litmus litmus) {
     if (!areRowTypesEqual(type1, type2, false)) {
-      return litmus.fail("Type mismatch:\n{}:\n{}\n{}:\n{}",
+      return litmus.fail("Type mismatch:\n{}:\n{}\n{}:\n{}\ndifference:\n{}",
           desc1, type1.getFullTypeString(),
-          desc2, type2.getFullTypeString());
+          desc2, type2.getFullTypeString(),
+          getFullTypeDifferenceString(type1, type2));
     }
     return litmus.succeed();
   }
 
+  public static String getFullTypeDifferenceString(RelDataType other,
+      RelDataType another) {
+    if (other == another) {
+      return "";
+    }
+
+    if (other.getFieldCount() != another.getFieldCount()) {
+      return "Type mismatch: the field size is not equal,"
+         + " one is " + other.getFieldCount()
+         + ", another one is " + another.getFieldCount();
+    }
+
+    final StringBuilder stringBuilder = new StringBuilder("Type mismatch: \n");
+    final List<RelDataTypeField> f1 = other.getFieldList();
+    final List<RelDataTypeField> f2 = another.getFieldList();
+    for (Pair<RelDataTypeField, RelDataTypeField> pair : Pair.zip(f1, f2)) {
+      final RelDataType type1 = pair.left.getType();
+      final RelDataType type2 = pair.right.getType();
+      // If one of the types is ANY comparison should succeed
+      if (type1.getSqlTypeName() == SqlTypeName.ANY
+          || type2.getSqlTypeName() == SqlTypeName.ANY) {
```

Comment:
Do you think the method could take a lambda for the equivalence rather than baking the logic here?
I think we should use reuse the comparison rather than duplicate logic.

Reply:
Do you mean this method is the same as `#areRowTypesEqual`?

Reply:
yes, the method is the same.

Reply:
They are a bit different. `areRowTypesEqual` would quit immediately if two fields are not equal, while `getFullTypeDifferenceString` would find all the different fields.

Reply:
I mean the comparison rules should better be reused rather than duplicated.

Reply:
I could not figure out a way to reuse those codes elegantly. Feel free to improve it if you have one~~